jito_searcher_client/async_searcher.py

```diff
@@ -0,0 +1,178 @@
+            await self.full_authentication()
+        elif now >= self._access_token.expiration:
+            await self.refresh_authentication()
+
+    async def refresh_authentication(self):
+        """
+        Performs an authentication refresh with the block engine, which involves using the refresh token to get a new
+        access token.
+        """
+        credentials = ssl_channel_credentials()
+        channel = secure_channel(self._url, credentials)
+        auth_client = AuthServiceStub(channel)
+
+        new_access_token: RefreshAccessTokenResponse = await auth_client.RefreshAccessToken(
+            RefreshAccessTokenRequest(refresh_token=self._refresh_token.token)
+        )
+        self._access_token = JwtToken(
+            token=new_access_token.access_token.value, expiration=new_access_token.access_token.expires_at_utc.seconds
+        )
+
+    async def full_authentication(self):
+        """
+        Performs full authentication with the block engine
+        """
+        credentials = ssl_channel_credentials()
+        channel = secure_channel(self._url, credentials)
+        auth_client = AuthServiceStub(channel)
+
+        challenge = (
+            await auth_client.GenerateAuthChallenge(
+                GenerateAuthChallengeRequest(role=Role.SEARCHER, pubkey=bytes(self._kp.pubkey()))
+            )
+        ).challenge
+
+        challenge_to_sign = f"{str(self._kp.pubkey())}-{challenge}"
+
+        signed = self._kp.sign_message(bytes(challenge_to_sign, "utf8"))
+
+        auth_tokens_response: GenerateAuthTokensResponse = await auth_client.GenerateAuthTokens(
+            GenerateAuthTokensRequest(
+                challenge=challenge_to_sign,
+                client_pubkey=bytes(self._kp.pubkey()),
+                signed_challenge=bytes(signed),
+            )
+        )
+
+        self._access_token = JwtToken(
+            token=auth_tokens_response.access_token.value,
+            expiration=auth_tokens_response.access_token.expires_at_utc.seconds,
+        )
+
+        self._refresh_token = JwtToken(
+            token=auth_tokens_response.refresh_token.value,
+            expiration=auth_tokens_response.refresh_token.expires_at_utc.seconds,
+        )
+
+
+async def get_async_searcher_client(url: str, kp: Keypair) -> SearcherServiceStub:
+    """
+    Returns a Searcher Service client that intercepts requests and authenticates with the block engine.
+    :param url: url of the block engine without http/https
+    :param kp: keypair of the block engine
+    :return: SearcherServiceStub which handles authentication on requests
+    """
+    # Authenticate immediately
+    searcher_interceptor = AsyncSearcherInterceptor(url, kp)
+    await searcher_interceptor.authenticate_if_needed()
+
+    credentials = ssl_channel_credentials()
+    channel = secure_channel(url, credentials, interceptors=[searcher_interceptor])
```

Comment:
There is currently an open [issue](https://github.com/grpc/grpc/issues/31442) with interceptors for grpcio. A temporary workaround could be to do something like this
```suggestion
    channel = secure_channel(url, credentials, interceptors=[searcher_interceptor])
    channel._unary_stream_interceptors.append(searcher_interceptor)
```